[JITLink] Add GOT indirection optimization for SystemZ

anoopkg6 opened the pull request:
Optimize the GOT and Stub relocations if the edge target address is in range from call site - Indirect jump by plt stub can be replaced with direct jump to target in post-allocation optimization.

Commit:
[JITLink] Add Got indirection optimization

Commit:
Merge branch 'main' of github.com:anoopkg6/llvm-project into opt-jitlink

anoopkg6 commented:
@uweigand

uweigand commented on llvm/lib/ExecutionEngine/JITLink/systemz.cpp:
```diff
@@ -106,11 +106,51 @@ const char *getEdgeKindName(Edge::Kind R) {
+        orc::ExecutorAddr TargetAddr = GOTTarget.getAddress();
+
+        int64_t Displacement = TargetAddr - EdgeAddr;
+        if (isInt<32>(Displacement)) {
```
Also, this should be `isInt<33>` here.

uweigand reviewed (commented):
Looks generally good, but see inline comments.  Thanks!

anoopkg6 committed:
Merge branch 'llvm:main' into opt-jitlink

anoopkg6 commented on llvm/include/llvm/ExecutionEngine/JITLink/systemz.h:
```diff
@@ -219,6 +219,24 @@ enum EdgeKind_systemz : Edge::Kind {
+  ///   - The result of the fixup expression  before shifting right by 1 must
+  ///     be multiple of 2, otherwise an alignment error will be returned.
+  ///
+  Delta32dblToPtrJumpStubBypassable,
```
I think we need a distinct name for DeltraPLT32dbl in PLT as there is RequestGOTAndTransformToDelta32dbl.

uweigand replied:
I do not understand this.   Whether the stub has or has not already been created only depends on the ordering of the passes - as long as the PLT pass happens before the GOT optimization pass, they will have been created; if the PLT pass happens too late they will not.  This is true whether or not the fixup kind is renamed inside the PLT pass.  What's the point of renaming the kind first in the PLT pass and then later eliminate it in the GOT optimizer pass?

Commit:
Added addend to edge offset from target qnd check isInt<33> to optimize GOT/Stub

anoopkg6 committed:
Merge branch 'llvm:main' into opt-jitlink

anoopkg6 committed:
Merge branch 'llvm:main' into opt-jitlink

Commit:
Merge branch 'opt-jitlink' of github.com:anoopkg6/llvm-project into opt-jitlink

anoopkg6 committed:
Merge branch 'llvm:main' into opt-jitlink

Commit:
Merge branch 'opt-jitlink' of github.com:anoopkg6/llvm-project into opt-jitlink

anoopkg6 committed:
Merge branch 'llvm:main' into opt-jitlink

anoopkg6 committed:
Merge branch 'llvm:main' into opt-jitlink

Commit:
1. Don't rename DeltaPLT32dbl to Delta32dblToPtrJumpStubBypassable in PLT pass.
2. Fix the issue with wrongly setting edge kind systemz::DeltaPLT32dbl for
   systemz::RequestGOTAndTransformToDelta32dbl instead of systemz::Delta32dbl.

anoopkg6 committed:
Merge branch 'llvm:main' into opt-jitlink

Commit:
Merge branch 'opt-jitlink' of github.com:anoopkg6/llvm-project into opt-jitlink

uweigand reviewed (approved):
LGTM now, thanks!
